topology.go

```diff
@@ -53,32 +92,144 @@ func (r *Reader) findLatestGeneration() (*generation, error) {
 		consistency = gocql.All
 	}
 
-	iter := r.config.Session.Query(queryString).Consistency(consistency).Iter()
+	gf := &generationFetcher{
+		session:        session,
+		clusterTracker: clusterTracker,
+		genTableName:   tableName,
+		lastTime:       startFrom,
+		logger:         logger,
+
+		consistency: consistency,
+
+		generationCh: make(chan *generation, 1),
+		stopCh:       make(chan struct{}),
+		refreshCh:    make(chan struct{}, 1),
+	}
+	return gf, nil
+}
+
+func (gf *generationFetcher) Run(ctx context.Context) error {
+	l := gf.logger
+
+	queryGensUpToTime := gf.session.Query(fmt.Sprintf("SELECT time, streams FROM %s WHERE time <= ? ALLOW FILTERING", gf.genTableName))
+	queryGensAfterTime := gf.session.Query(fmt.Sprintf("SELECT time, streams FROM %s WHERE time > ? ALLOW FILTERING", gf.genTableName))
+
+	// Find the generation which contains the timestamp from which we should
+	// start reading
+	gl, err := gf.fetchFromGenerationsTable(queryGensUpToTime, gf.lastTime)
+	if err != nil {
+		close(gf.generationCh)
+		return err
+	}
+
+	if len(gl) > 0 {
+		first := gl[len(gl)-1]
+
+		gf.logger.Printf("pushing generation %v", first.startTime)
+		gf.generationCh <- first
+		gf.lastTime = first.startTime
+	}
+
+	// Periodically poll for newer generations
+	l.Printf("starting generation fetcher loop")
+outer:
+	for {
+		// Generation processing can take some time, so start calculating
+		// the next poll time starting from now
+		waitC := time.After(generationFetchPeriod)
 
-	var timestamp, bestTimestamp, expired time.Time
-	var streams, bestStreams []stream
+		// See if there are any new generations
+		gl, err := gf.fetchFromGenerationsTable(queryGensAfterTime, gf.lastTime)
```

Comment:
The periodic fetch should only query `time`, and only if we find that there is a new `time`, query its `streams` (using `WHERE ... =` query). Fetching `streams` on each periodic query is very wasteful

Reply:
I reimplemented the `generationFetcher`. Now, it polls for generation times only, without using `ALLOW FILTERING`, and fetches streams for a generation only if necessary.